### PR TITLE
Feature/online sessions

### PR DESCRIPTION
**Pull request by TheEmidee**

### CODE REVIEW

File: Source/GameBaseFramework/Private/BlueprintLibraries/GBFHelperBlueprintLibrary.cpp [hunk 1]
```diff
@@ -1,9 +1,11 @@
 #include "BlueprintLibraries/GBFHelperBlueprintLibrary.h"
 
+#include "Engine/Engine.h"
```
Comment (WannesVanHooste) [inline Source/GameBaseFramework/Private/BlueprintLibraries/GBFHelperBlueprintLibrary.cpp:3]: <>

File: Source/GameBaseFramework/Private/BlueprintLibraries/GBFHelperBlueprintLibrary.cpp [hunk 2]
```diff
@@ -17,4 +19,31 @@ void UGBFHelperBlueprintLibrary::OpenMap( const UObject * world_context, const T
     }
 
     UGameplayStatics::OpenLevel( world_context, FName( *map_soft_object_ptr.GetAssetName() ), true );
-}
\ No newline at end of file
+}
+
+bool UGBFHelperBlueprintLibrary::BrowseMap( FWorldContext & world_context, const TSoftObjectPtr< UWorld > & map_soft_object_ptr, const bool open_if_current /* = false */ )
```
Comment (WannesVanHooste) [inline Source/GameBaseFramework/Private/BlueprintLibraries/GBFHelperBlueprintLibrary.cpp:24]: remove comment

File: Source/GameBaseFramework/Private/Engine/SubSystems/GBFGameInstanceIdentitySubsystem.cpp [hunk 3]
```diff
@@ -66,21 +73,181 @@ bool UGBFGameInstanceIdentitySubsystem::ShowLoginUI( const int controller_index,
...
+                EGBFUIDialogType::AdditiveOnTop );
+        }
+
+        return false;
+    }
+
+    return true;
+}
+
 void UGBFGameInstanceIdentitySubsystem::HandleUserLoginChanged( const int32 game_user_index, ELoginStatus::Type /* previous_login_status */, const ELoginStatus::Type login_status, const FUniqueNetId & user_id )
```
Comment (WannesVanHooste) [inline Source/GameBaseFramework/Private/Engine/SubSystems/GBFGameInstanceIdentitySubsystem.cpp:219]: remove comment

File: Source/GameBaseFramework/Private/Engine/SubSystems/GBFGameInstanceSessionSubsystem.cpp [hunk 4]
```diff
@@ -0,0 +1,782 @@
...
+bool UGBFGameInstanceSessionSubsystem::JoinSession( ULocalPlayer * local_player, const FOnlineSessionSearchResult & search_result )
+{
+    if ( auto * game_session = GetGameSession() )
+    {
+        AddNetworkFailureHandlers();
+
+        OnJoinSessionCompleteDelegateHandle = game_session->OnJoinSessionComplete().AddUObject( this, &UGBFGameInstanceSessionSubsystem::OnJoinSessionComplete );
+        if ( game_session->JoinSession( local_player->GetPreferredUniqueNetId().GetUniqueNetId(), NAME_GameSession, search_result ) )
+        {
+            // If any error occured in the above, pending state would be set
```
Comment (WannesVanHooste) [inline Source/GameBaseFramework/Private/Engine/SubSystems/GBFGameInstanceSessionSubsystem.cpp:87]: remove if not meaningfull

File: Source/GameBaseFramework/Public/Engine/SubSystems/GBFGameInstanceSessionSubsystem.h [hunk 5]
```diff
@@ -0,0 +1,148 @@
+#pragma once
+
+#include "Engine/Engine.h"
+#include "GBFGameInstanceSubsystemBase.h"
+#include "Interfaces/OnlineIdentityInterface.h"
+#include "Online/GBFGameSession.h"
+#include "OnlineSessionSettings.h"
```
Comment (VectorGamez) [inline Source/GameBaseFramework/Public/Engine/SubSystems/GBFGameInstanceSessionSubsystem.h:7]: <>

File: Source/GameBaseFramework/Public/Engine/SubSystems/GBFGameInstanceSessionSubsystem.h [hunk 6]
```diff
@@ -0,0 +1,148 @@
+#pragma once
+
+#include "Engine/Engine.h"
```
Comment (VectorGamez) [inline Source/GameBaseFramework/Public/Engine/SubSystems/GBFGameInstanceSessionSubsystem.h:3]: <>

File: Source/GameBaseFramework/Public/Online/GBFGameSession.h [hunk 7]
```diff
@@ -0,0 +1,108 @@
+#pragma once
+
+#include "Templates/SharedPointerInternals.h"
+#include "Templates/SharedPointerInternals.h"
+#include "Templates/SharedPointerInternals.h"
+#include "Templates/SharedPointerInternals.h"
+#include "Templates/SharedPointerInternals.h"
```
Comment (VectorGamez) [inline Source/GameBaseFramework/Public/Online/GBFGameSession.h:7]: <>
Should only be once

File: Source/GameBaseFramework/Private/Engine/SubSystems/GBFGameInstanceSessionSubsystem.cpp [hunk 8]
```diff
@@ -0,0 +1,782 @@
...
+        {
+            UE_LOG( LogOnline, Log, TEXT( "Waiting for session %s to start, and then we will end it to return to main menu" ), *game_session.ToString() );
+            OnStartSessionCompleteDelegateHandle = sessions->AddOnStartSessionCompleteDelegate_Handle( OnEndSessionCompleteDelegate );
+            pending_online_operation = true;
+        }
+    }
+
+    if ( !pending_online_operation )
+    {
+        //GEngine->HandleDisconnect( GetWorld(), GetWorld()->GetNetDriver() );
```
Comment (WannesVanHooste) [inline Source/GameBaseFramework/Private/Engine/SubSystems/GBFGameInstanceSessionSubsystem.cpp:405]: remove if unnessary

File: Source/GameBaseFramework/Private/Engine/SubSystems/GBFGameInstanceSessionSubsystem.cpp [hunk 9]
```diff
@@ -0,0 +1,782 @@
...
+        }
+    }
+
+    if ( !error_reason.IsEmpty() )
+    {
+        RemoveNetworkFailureHandlers();
+        GetGBFGameInstance()->ShowMessageThenGotoMainMenuState( NSLOCTEXT( "NetworkErrors", "LocKey_JoinSessionFailedTitle", "Impossible to join the session" ), error_reason );
+        return;
+    }
+    // Add debug encryption token if desired.
```
Comment (WannesVanHooste) [inline Source/GameBaseFramework/Private/Engine/SubSystems/GBFGameInstanceSessionSubsystem.cpp:681]: remove comments if unnessary

File: Source/GameBaseFramework/Private/Online/GBFGameSession.cpp [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+    if ( auto * oss = Online::GetSubsystem( GetWorld() ) )
+    {
+        auto session_interface_ptr = oss->GetSessionInterface();
+        session_interface_ptr->ClearOnCreateSessionCompleteDelegate_Handle( OnCreateSessionCompleteDelegateHandle );
+    }
+
+    OnCreatePresenceSessionCompleteEvent.Broadcast( session_name, was_successful );
+}
+
+void AGBFGameSession::OnStartOnlineGameComplete( FName /* session_name */, const bool was_successful )
```
Comment (WannesVanHooste) [inline Source/GameBaseFramework/Private/Online/GBFGameSession.cpp:236]: no argument in function?

File: Source/GameBaseFramework/Private/Online/GBFOnlineSessionClient.cpp [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+        return;
+    }
+
+    if ( !user_id.IsValid() )
+    {
+        UE_LOG( LogOnline, Warning, TEXT( "Invite accept returned no user." ) );
+        return;
+    }
+
+    if ( auto * game_instance = Cast< UGBFGameInstance >( GetGameInstance() ) )
```
Comment (WannesVanHooste) [inline Source/GameBaseFramework/Private/Online/GBFOnlineSessionClient.cpp:27]: setting variable in if statement?

File: Source/GameBaseFramework/Public/Engine/GBFGameInstance.h [hunk 12]
```diff
@@ -1,14 +1,16 @@
 #pragma once
 
-#include <Containers/Ticker.h>
+#include "Containers/BackgroundableTicker.h"
```
Comment (WannesVanHooste) [inline Source/GameBaseFramework/Public/Engine/GBFGameInstance.h:3]: <> if this is not one of your classes

File: Source/GameBaseFramework/Public/Engine/SubSystems/GBFGameInstanceSessionSubsystem.h (hunk 6 above)
Comment (WannesVanHooste) [inline Source/GameBaseFramework/Public/Engine/SubSystems/GBFGameInstanceSessionSubsystem.h:3]: <>

File: Source/GameBaseFramework/Public/Engine/SubSystems/GBFGameInstanceSessionSubsystem.h [hunk 13]
```diff
@@ -0,0 +1,148 @@
+#pragma once
+
+#include "Engine/Engine.h"
+#include "GBFGameInstanceSubsystemBase.h"
+#include "Interfaces/OnlineIdentityInterface.h"
```
Comment (WannesVanHooste) [inline Source/GameBaseFramework/Public/Engine/SubSystems/GBFGameInstanceSessionSubsystem.h:5]: <> if not one of your classes

File: Source/GameBaseFramework/Public/Engine/SubSystems/GBFGameInstanceSessionSubsystem.h (hunk 5 above)
Comment (WannesVanHooste) [inline Source/GameBaseFramework/Public/Engine/SubSystems/GBFGameInstanceSessionSubsystem.h:7]: <> if not one of your classes

File: Source/GameBaseFramework/Private/Engine/SubSystems/GBFGameInstanceIdentitySubsystem.cpp (hunk 3 above)
Comment (TheEmidee) [reply Source/GameBaseFramework/Private/Engine/SubSystems/GBFGameInstanceIdentitySubsystem.cpp:219]: cpp.function.args.unused

File: Source/GameBaseFramework/Private/BlueprintLibraries/GBFHelperBlueprintLibrary.cpp (hunk 2 above)
Comment (TheEmidee) [reply Source/GameBaseFramework/Private/BlueprintLibraries/GBFHelperBlueprintLibrary.cpp:24]: cpp.function.args.unused

File: Source/GameBaseFramework/Private/Online/GBFGameSession.cpp (hunk 10 above)
Comment (TheEmidee) [reply Source/GameBaseFramework/Private/Online/GBFGameSession.cpp:236]: yes there was_successful

File: Source/GameBaseFramework/Private/Engine/SubSystems/GBFGameInstanceSessionSubsystem.cpp (hunk 9 above)
Comment (TheEmidee) [reply Source/GameBaseFramework/Private/Engine/SubSystems/GBFGameInstanceSessionSubsystem.cpp:681]: done

File: Source/GameBaseFramework/Private/Engine/SubSystems/GBFGameInstanceSessionSubsystem.cpp (hunk 8 above)
Comment (TheEmidee) [reply Source/GameBaseFramework/Private/Engine/SubSystems/GBFGameInstanceSessionSubsystem.cpp:405]: done

File: Source/GameBaseFramework/Public/Online/GBFGameSession.h (hunk 7 above)
Comment (TheEmidee) [reply Source/GameBaseFramework/Public/Online/GBFGameSession.h:7]: done

File: Source/GameBaseFramework/Public/Engine/SubSystems/GBFGameInstanceSessionSubsystem.h (hunk 5 above)
Comment (TheEmidee) [reply Source/GameBaseFramework/Public/Engine/SubSystems/GBFGameInstanceSessionSubsystem.h:7]: done

File: Source/GameBaseFramework/Private/Engine/SubSystems/GBFGameInstanceSessionSubsystem.cpp (hunk 4 above)
Comment (TheEmidee) [reply Source/GameBaseFramework/Private/Engine/SubSystems/GBFGameInstanceSessionSubsystem.cpp:87]: done

File: Source/GameBaseFramework/Private/BlueprintLibraries/GBFHelperBlueprintLibrary.cpp (hunk 1 above)
Comment (TheEmidee) [reply Source/GameBaseFramework/Private/BlueprintLibraries/GBFHelperBlueprintLibrary.cpp:3]: done

File: Source/GameBaseFramework/Public/Engine/GBFGameInstance.h (hunk 12 above)
Comment (TheEmidee) [reply Source/GameBaseFramework/Public/Engine/GBFGameInstance.h:3]: done

File: Source/GameBaseFramework/Public/Engine/SubSystems/GBFGameInstanceSessionSubsystem.h (hunk 6 above)
Comment (TheEmidee) [reply Source/GameBaseFramework/Public/Engine/SubSystems/GBFGameInstanceSessionSubsystem.h:3]: done

File: Source/GameBaseFramework/Public/Engine/SubSystems/GBFGameInstanceSessionSubsystem.h (hunk 13 above)
Comment (TheEmidee) [reply Source/GameBaseFramework/Public/Engine/SubSystems/GBFGameInstanceSessionSubsystem.h:5]: done

**Pull request merged**